official/recommendation/ncf_dataset.py

```diff
@@ -0,0 +1,170 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Prepare dataset for NCF.
+
+Load the training dataset and evaluation dataset from csv file into memory.
+Prepare input for model training and evaluation.
+"""
+import numpy as np
+import scipy.sparse as sp
```

Comment:
This is another dependency-- note that it will have to be added, along with pandas, to the list of dependencies, and also to any BUILD files.

Reply:
On second thought, starting to review this file, but I think we do this in TF instead of numpy/scipy. That would likely improve performance, and reduce dependencies. See https://www.tensorflow.org/get_started/datasets_quickstart#reading_a_csv_file , and then the rest should just be general matrix manipulation, which should be done in TF directly. Not going to review the rest of this file now, but will re-review when done as "TF-first".

Reply:
Okay, will do.